tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java

```diff
@@ -0,0 +1,103 @@
+
+import org.apache.maven.SessionScoped;
+import org.apache.maven.artifact.ArtifactUtils;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.model.Model;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.repository.internal.MavenWorkspaceReader;
+import org.codehaus.plexus.component.annotations.Component;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.repository.WorkspaceReader;
+import org.eclipse.aether.repository.WorkspaceRepository;
+import org.eclipse.tycho.PackagingType;
+
+/**
+ * An implementation of a workspace reader similar to what maven does but for the Tycho packaging
+ * types
+ */
+@SessionScoped
+@Component(role = WorkspaceReader.class, hint = TychoReactorReader.HINT)
+public class TychoReactorReader implements MavenWorkspaceReader {
+    static final String HINT = "tycho-reactor";
+
+    private final Map<String, MavenProject> projectsByGAV;
+    private final WorkspaceRepository repository;
+
+    @Inject
+    public TychoReactorReader(MavenSession session) {
+        this.projectsByGAV = session.getProjects().stream()
+                .collect(toMap(s -> ArtifactUtils.key(s.getGroupId(), s.getArtifactId(), s.getVersion()), identity()));
+        repository = new WorkspaceRepository(HINT, null);
+    }
+
+    @Override
+    public WorkspaceRepository getRepository() {
+        return repository;
+    }
+
+    @Override
+    public File findArtifact(Artifact artifact) {
+        return getTychoReactorProject(artifact).map(project -> {
+            org.apache.maven.artifact.Artifact mainArtifact = project.getArtifact();
+            if (mainArtifact != null && mainArtifact.getFile() != null && mainArtifact.getFile().exists()) {
+                return mainArtifact.getFile();
+            }
+            return project.getBasedir();
+        }).orElse(null);
+    }
+
+    private Optional<MavenProject> getTychoReactorProject(Artifact artifact) {
+        if (isTychoReactorArtifact(artifact)) {
+            String projectKey = ArtifactUtils.key(artifact.getGroupId(), artifact.getArtifactId(),
+                    artifact.getVersion());
+            MavenProject project = projectsByGAV.get(projectKey);
+            if (project != null) {
+                if (PackagingType.TYCHO_PACKAGING_TYPES.contains(project.getPackaging())) {
+                    return Optional.of(project);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    private boolean isTychoReactorArtifact(Artifact artifact) {
+        if (artifact.getClassifier() == null || artifact.getClassifier().isBlank()) {
+            return PackagingType.TYCHO_PACKAGING_TYPES.contains(artifact.getProperty("type", ""));
+        }
+        return false;
+    }
+
+    @Override
+    public List<String> findVersions(Artifact artifact) {
+        return getTychoReactorProject(artifact).map(project -> List.of(artifact.getVersion())).orElse(List.of());
```

Comment:
In my custom extension I always return an empty list here. 
If I recall correctly this is basically unused.

Reply:
`DefaultVersionRangeResolver` calls this method.

Reply:
@laeubi ahh yes, you're right, not sure why the search didn't bring up any reference.